[CORE-8669] `storage`: early abort compaction

WillemKauf opened the pull request:
### Converting to draft, no need to review. We will work on separating garbage collection and compaction in separate loops first, and then worry about the ideas here.


Compaction can be a long running process, during which prefix truncation and regularly scheduled garbage collection is blocked.

This PR, most notably, adds `compaction_config::maybe_abort_compaction()` which checks the abort source (as was done before) but also now considers two signals for terminating the compaction process early:

1. Degraded or low disk space alerts
2. Resource management triggered garbage collection

In either of these cases, compaction is aborted early.

## Backports Required

- [X] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes

### Improvements

* Allow the early abortion of compaction in the cases of disk pressure or resource management triggered garbage collection in order to prevent long running compactions from blocking emergency retention measures

WillemKauf committed:
`storage`: add `gc_required_exception`

This exception is intended to be thrown in the case of early
compaction abortion.

WillemKauf committed:
`storage`: add `disk_log_impl::has_cloud_gc_offset()`

WillemKauf committed:
`storage`: add `log_manager::gc_required()`

Signals that garbage collection is required ahead of or in the middle
of compaction. This returns `true` in low or degraded disk space scenarios,
as well as regularly triggered garbage collection.

WillemKauf committed:
`storage`: add `maybe_abort_compaction()` to `compaction_config`

Also add member variables for `disk_space_alert` and `disk_log_impl`
for use in determining whether compaction should be aborted.

WillemKauf committed:
`storage`: plumb `maybe_abort_compaction()` into compaction utils

WillemKauf committed:
`storage`: add early abort compaction tests

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--a8f8fd42-5179-4cf4-87dc-8d4501c53e7a-->
<!--TEST RESULTS-->
<!--08c5dfd3-72f3-404b-9a5f-5a2fa431ef4b-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/61850">build#61850</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| kafka_server_rpfixture.kafka_server_rpfixture | unit | https://buildkite.com/redpanda/redpanda/builds/61850#01950115-5ef3-4b31-bb8d-1fe7282fddef | FLAKY | 1/2 |
| rptest.tests.compaction_recovery_test.CompactionRecoveryTest.test_index_recovery | ducktape | https://buildkite.com/redpanda/redpanda/builds/61850#0195015e-543e-4801-a743-7fc4e6818184 | FLAKY | 1/2 |
| rptest.tests.log_compaction_test.LogCompactionTest.compaction_stress_test.cleanup_policy=compact.delete.key_set_cardinality=1000.storage_compaction_key_map_memory_kb=10 | ducktape | https://buildkite.com/redpanda/redpanda/builds/61850#0195015e-543e-4801-a743-7fc4e6818184 | FLAKY | 1/2 |
| rptest.tests.partition_force_reconfiguration_test.NodeWiseRecoveryTest.test_node_wise_recovery.dead_node_count=1 | ducktape | https://buildkite.com/redpanda/redpanda/builds/61850#0195015a-2370-45b5-a920-0a174cea0d6b | FLAKY | 1/2 |
| rptest.tests.partition_move_interruption_test.PartitionMoveInterruption.test_forced_cancellation | ducktape | https://buildkite.com/redpanda/redpanda/builds/61850#0195015a-2370-4e70-9459-a943a5ae100a | FLAKY | 1/2 |
| rptest.tests.write_caching_fi_e2e_test.WriteCachingFailureInjectionE2ETest.test_crash_all_with_consumer_group | ducktape | https://buildkite.com/redpanda/redpanda/builds/61850#0195015a-2370-45b5-a920-0a174cea0d6b | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/61886">build#61886</a></summary>


| test_id                                                                                                                                                                 | test_kind   | job_url                                                                                   | test_status   | passed   |
|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.compaction_recovery_test.CompactionRecoveryTest.test_index_recovery                                                                                        | ducktape    | https://buildkite.com/redpanda/redpanda/builds/61886#01950820-a6a9-4b1b-9ae7-8dd9907650c7 | FLAKY         | 1/5      |
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade                                                                   | ducktape    | https://buildkite.com/redpanda/redpanda/builds/61886#01950820-a6aa-4dc9-9294-b655c5cbbe23 | FLAKY         | 1/2      |
| rptest.tests.log_compaction_test.LogCompactionTest.compaction_stress_test.cleanup_policy=compact.delete.key_set_cardinality=1000.storage_compaction_key_map_memory_kb=3 | ducktape    | https://buildkite.com/redpanda/redpanda/builds/61886#0195083b-4038-4ee7-88b5-088f069e1136 | FLAKY         | 1/2      |
| rptest.tests.log_compaction_test.LogCompactionTest.compaction_stress_test.cleanup_policy=compact.key_set_cardinality=1000.storage_compaction_key_map_memory_kb=3        | ducktape    | https://buildkite.com/redpanda/redpanda/builds/61886#0195083b-4039-4f96-add1-0040eb874393 | FLAKY         | 1/2      |
</details>
<!--/body-->

Lazin reviewed (commented):
how does it interact with space management?

WillemKauf commented:
>how does it interact with space management?

Thanks for asking.

For a tiered-storage (archival) enabled topic, in [`disk_space_manager::manage_data_disk`](https://github.com/redpanda-data/redpanda/blob/59ee571fe90251804f1e53afd1857b9466f5df10/src/v/resource_mgmt/storage.cc#L509) we use the current disk usage and retention configurations to install an eviction schedule, the severity of which is dependent on how much we are able to reclaim at each level (i.e up to the configured local retention limits, low space level, active segment).

This ultimately has the effect of [setting `cloud_gc_offset` in the log to an offset per the eviction schedule](https://github.com/redpanda-data/redpanda/blob/59ee571fe90251804f1e53afd1857b9466f5df10/src/v/resource_mgmt/storage.cc#L346), to be used and reset later in [`disk_log_impl::do_gc()`](https://github.com/redpanda-data/redpanda/blob/59ee571fe90251804f1e53afd1857b9466f5df10/src/v/storage/disk_log_impl.cc#L1521-L1530) as an override for the eviction offset (which then feeds into the `log_eviction_stm` to perform the prefix truncation up to that point).

With this PR, we now (repeatedly) check if the log being compacted has a `cloud_gc_offset` throughout the compaction process, and will abort compaction early if it does (as this is an indicator that resource management has requested eviction in order to free up disk space, and we should bail out of compaction to allow the prefix truncation to occur).

Let me know if this clears it up.

WillemKauf committed:
`storage`: add new check for aborting compaction

All of the current checks to `maybe_abort_compaction()` happen after
some work has already been done in the compaction process.

Add an extra check near the start of compaction to early return
as soon as possible in the presence of disk pressure or space
management.

WillemKauf commented:
Push to:
* Add an extra check near the start of compaction to early return as soon as possible in the presence of disk pressure or space management.

andrwng commented on src/v/storage/types.cc:
```diff
@@ -244,4 +246,35 @@ operator<<(std::ostream& o, compacted_index::recovery_state state) {
+            }
+        }
+
+        // If resource management has set a cloud_gc_offset, bail out of
+        // compaction early to allow prefix truncation to quickly reclaim data.
+        if (log.has_cloud_gc_offset()) {
+            throw gc_required_exception(
+              fmt::format("Bailing out of compaction due to resource "
+                          "management setting cloud_gc_offset"));
+        }
```
Hmm isn't the cloud GC offset almost always set when there's a stream of incoming data in the steady state? I thought typically space management lets local storage grow, but at around 70-80% it tries to reclaim from all partitions every 30 seconds?

Or is the idea that compaction always takes under 30 seconds, so there's no risk of starving compactions?

WillemKauf replied:
> Hmm isn't the cloud GC offset almost always set when there's a stream of incoming data in the steady state? I thought typically space management lets local storage grow, but at around 70-80% it tries to reclaim from all partitions every 30 seconds?

@dotnwat and I were discussing this potential, highly undesirable state, in which compaction is constantly being bailed out of because the cloud GC offset is set, but space management is not reclaiming data for whatever reason/or not reclaiming enough data to get us out of the retention limit. All of this leading to 0 retention progress being made.

I'm not sure there's a great solution here yet, I will have to think about it more.

WillemKauf commented:
As a note here: the storage team has been brainstorming ways to run garbage collection and compaction concurrently in separate fibres, which would render this work _mostly_ unnecessary.
